Modules/Loadable/Markups/MRML/vtkMRMLMarkupsCurveNode.cxx

```diff
@@ -183,6 +189,205 @@ double vtkMRMLMarkupsCurveNode::GetCurveLengthBetweenStartEndPointsWorld(vtkIdTy
+		double startNormal[3] = { 0.0 };
+		normalVectorArray->GetTuple(pointIdStart, startNormal);
+		vtkIdType pointIdEnd = pointLocator->FindClosestPoint(segmentEndPoint);
+		double endNormal[3] = { 0.0 };
+		normalVectorArray->GetTuple(pointIdStart, endNormal);
+		double rayDirection[3] = { 0.0 };
+		rayDirection[0] = (startNormal[0] + endNormal[0]) / 2;
+		rayDirection[1] = (startNormal[1] + endNormal[1]) / 2;
+		rayDirection[2] = (startNormal[2] + endNormal[2]) / 2;
+		pointNormalArray->InsertNextPoint(rayDirection);
+		
+	}
+	vtkMRMLMarkupsCurveNode::ConstrainPointsToSurface(interpolatedPoints, pointNormalArray, surfacePolydata, snappedToSurfacePoints);
+
+	this->SetControlPointPositionsWorld(snappedToSurfacePoints);
+	this->SetControlPointLabelsWorld(originalLabels, originalControlPoints);
+}
+
+//---------------------------------------------------------------------------
+vtkIdType vtkMRMLMarkupsCurveNode::GetClosestControlPointIndexToPositionWorld(const double posWorld[3])
+{
+	vtkNew<vtkPoints> controlPoints;
+	this->GetControlPointPositionsWorld(controlPoints);
+
+	if (!controlPoints)
+	{
+		return -1;
+	}
+
+	vtkIdType closestControlPointID = 0;
+	vtkIdType numberOfPoints = controlPoints->GetNumberOfPoints();
+	double closestPointDistance = vtkMath::Distance2BetweenPoints(posWorld, controlPoints->GetPoint(0));
+
+	for (vtkIdType pointIndex = 1; pointIndex < numberOfPoints; pointIndex++)
+	{
+		double* nextPoint = controlPoints->GetPoint(pointIndex);
+		double nextPointDistance = vtkMath::Distance2BetweenPoints(posWorld, nextPoint);
+		if (nextPointDistance < closestPointDistance)
+		{
+			closestPointDistance = nextPointDistance;
+			closestControlPointID = pointIndex;
+		}
+	}
+
+	return closestControlPointID;
+}
+//---------------------------------------------------------------------------
+bool vtkMRMLMarkupsCurveNode::ConstrainPointsToSurface(vtkPoints* originalPoints, vtkPoints* normalVectors, vtkPolyData* surfacePolydata, vtkPoints* surfacePoints)
+{
+	if (originalPoints->GetNumberOfPoints()!= normalVectors->GetNumberOfPoints())
+	{
+		vtkGenericWarningMacro("vtkMRMLMarkupsCurveNode::ResamplePoints failed: invalid inputs");
+		return false;
+	}
+	
+	vtkNew<vtkOBBTree> obbTree;
+	obbTree->SetDataSet(surfacePolydata);
+	obbTree->BuildLocator();
+
+	vtkNew<vtkPointLocator> pointLocator;
+	pointLocator->SetDataSet(surfacePolydata);
+	pointLocator->BuildLocator();
+
+	double originalPoint[3] = { 0.0 };
+	double rayDirection[3] = { 0.0 };
+	double exteriorPoint[3] = { 0.0 };
+
+	// Curve is expected to be close to surface. If projection distance is greater than 10% of image size
+	// it's likely due to a second external surface
+	double bounds[6] = { 0.0 };
+	surfacePolydata->GetBounds(bounds);
+	double point1[3] = { bounds[0], bounds[2], bounds[4]};
+	double point2[3] = { bounds[1], bounds[3], bounds[5] };
+	double polydataSize = vtkMath::Distance2BetweenPoints(point1,point2);
```

Comment:
It would be more elegant to use vtkBoundingBox::GetDiagonalLength to compute polydataSize. Also, size typically refers to integer count, so rename it something like length or diameter.